ompi/communicator/comm.c

```diff
@@ -1010,130 +1143,73 @@ int ompi_comm_split_type (ompi_communicator_t *comm, int split_type, int key,
-    /* --------------------------------------------------------- */
-    /* Create the communicator finally */
-
-    do {
-        rc = ompi_comm_set (&newcomp, comm, my_size, lranks, my_rsize,
-                            rranks, NULL, comm->error_handler, NULL, NULL, 0);
-        if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-            break;
-        }
-
-        /* Determine context id. It is identical to f_2_c_handle */
-        rc = ompi_comm_nextcid (newcomp, comm, NULL, NULL, NULL, false, mode);
-        if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
-            break;
-        }
+    if (MPI_COMM_TYPE_HW_UNGUIDED == global_orig_split_type) {
+        /* Handle MPI_COMM_TYPE_HW_UNGUIDED communicator split. */
+        return ompi_comm_split_unguided( comm, info, newcomm );
+    } else {
+        return ompi_comm_split_type_core( comm, inter, need_split, global_split_type,
+                                          split_type, key, info, no_reorder,
+                                          no_undefined, newcomm);
+    }
+}
 
-        ompi_comm_assert_subscribe (newcomp, OMPI_COMM_ASSERT_LAZY_BARRIER);
-        ompi_comm_assert_subscribe (newcomp, OMPI_COMM_ASSERT_ACTIVE_POLL);
-        if (info) {
-            opal_infosubscribe_change_info(&newcomp->super, info);
-        }
+int ompi_comm_split_unguided(ompi_communicator_t *comm, struct opal_info_t *info,
+                             ompi_communicator_t** newcomm) {
+    int rc;
+    struct opal_info_t *split_info = OBJ_NEW(opal_info_t);
+    int original_size = ompi_comm_size(comm);
+    int i;
 
-        /* Activate the communicator and init coll-component */
-        rc = ompi_comm_activate (&newcomp, comm, NULL, NULL, NULL, false, mode);
-        if (OPAL_UNLIKELY(OMPI_SUCCESS != rc)) {
+    if (1 == original_size) {
+        /* Communicator cannot be split any smaller */
+        *newcomm = MPI_COMM_NULL;
+        return OMPI_SUCCESS;
+    }
+    /*
+     * Attempt to split the communicator based on topology class by iteratively
+     * calling ompi_comm_split_type specifying the split type as
+     * MPI_COMM_TYPE_HW_GUIDED using the next lower topology class until a
+     * split results in a smaller size communicator than the input communicator.
+     * The search starts with OMPI_COMM_TYPE_CU since that is the highest possible
+     * topology class where the communicator size can be smaller than MPI_COMM_WORLD.
+     */
+    i = 1;
+    while (NULL != ompi_comm_split_type_hw_guided_support[i].info_value) {
+        int new_size;
+        /* MPI_COMM_TYPE_HW_GUIDED splits require mpi_hw_resource_type to be set */
+        opal_info_set(split_info, "mpi_hw_resource_type",
+                      ompi_comm_split_type_hw_guided_support[i].info_value);
+        rc = ompi_comm_split_type(comm, MPI_COMM_TYPE_HW_GUIDED,
+                                  ompi_comm_split_type_hw_guided_support[i].split_type,
+                                  split_info, newcomm);
```

Comment:
I think you should be calling `ompi_comm_split_type_core` instead of `ompi_comm_split_type` here. The MPI interface calls `ompi_comm_split_type` which then checks the user's parameters and calls `ompi_comm_split_unguided`. So there is no need to go through the user error checking again in this loop.

I'm thinking that is should look something like (you will need to pass in `inter`, `need_split `, `key`, `no_reorder`, and `no_undefined` from the `ompi_comm_split_type` function at the point of the call)
```suggestion
        rc = ompi_comm_split_type_core(comm, 
                                  MPI_COMM_TYPE_HW_GUIDED,
                                  ompi_comm_split_type_hw_guided_support[i].split_type,
                                  key, split_info,
                                  need_split, no_reorder, no_undefined,
                                  newcomm);
```

Since you already decoded the `MPI_COMM_TYPE_HW_GUIDED` into `ompi_comm_split_type_hw_guided_support[i].split_type` then you don't need to walk through that conversion again.

If the user passed `MPI_UNDEFINED` as their split type, you will likely need to know that here as well so you can pass it down to the `ompi_comm_split_type_core`. So you might need to add that to the function signature.